File: basic_dsa_five.java
/**
 * Basic DSA Problems - Two More Fundamental Questions
 * Simple and essential DSA problems for beginners
 */
public class basic_dsa_five {
    
    // Problem 1: Find Factorial of a Number
    // Calculate factorial using both iterative and recursive approaches
    
    /**
     * Iterative approach to find factorial
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static Long factorialIterative(int n) {
        if (n < 0) {
            return null; // Factorial is not defined for negative numbers
        }
        
        long result = 1;
        for (int i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }
    
    /**
     * Recursive approach to find factorial
     * Time Complexity: O(n)
     * Space Complexity: O(n) due to recursion stack
     */
    public static Long factorialRecursive(int n) {
        if (n < 0) {
            return null; // Factorial is not defined for negative numbers
        }
        if (n == 0 || n == 1) {
            return 1L;
        }
        return (long) n * factorialRecursive(n - 1);
    }
    
    // Problem 2: Check if a Number is Prime
    // Determine whether a number is prime or not
    
    /**
     * Check if a number is prime
     * Time Complexity: O(√n)
     * Space Complexity: O(1)
     */
    public static boolean isPrime(int n) {
        if (n < 2) {
            return false;
        }
        if (n == 2) {
            return true;
        }
        if (n % 2 == 0) {
            return false;
        }
        
        // Check divisibility from 3 to √n
        for (int i = 3; i * i <= n; i += 2) {
            if (n % i == 0) {
                return false;
            }
        }
        
        return true;
    }
    
    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Find Factorial of a Number");
        System.out.println("=".repeat(60));
        
        // Test Case 1
        int num1 = 5;
        Long factIter1 = factorialIterative(num1);
        Long factRec1 = factorialRecursive(num1);
        System.out.println("Number: " + num1);
        System.out.println("Factorial (Iterative): " + factIter1);
        System.out.println("Factorial (Recursive): " + factRec1);
        
        // Test Case 2
        int num2 = 0;
        Long factIter2 = factorialIterative(num2);
        Long factRec2 = factorialRecursive(num2);
        System.out.println("\nNumber: " + num2);
        System.out.println("Factorial (Iterative): " + factIter2);
        System.out.println("Factorial (Recursive): " + factRec2);
        
        // Test Case 3
        int num3 = 1;
        Long factIter3 = factorialIterative(num3);
        Long factRec3 = factorialRecursive(num3);
        System.out.println("\nNumber: " + num3);
        System.out.println("Factorial (Iterative): " + factIter3);
        System.out.println("Factorial (Recursive): " + factRec3);
        
        // Test Case 4
        int num4 = 7;
        Long factIter4 = factorialIterative(num4);
        Long factRec4 = factorialRecursive(num4);
        System.out.println("\nNumber: " + num4);
        System.out.println("Factorial (Iterative): " + factIter4);
        System.out.println("Factorial (Recursive): " + factRec4);
        
        // Test Case 5 - Negative number
        int num5 = -5;
        Long factIter5 = factorialIterative(num5);
        Long factRec5 = factorialRecursive(num5);
        System.out.println("\nNumber: " + num5);
        System.out.println("Factorial (Iterative): " + factIter5);
        System.out.println("Factorial (Recursive): " + factRec5);
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Problem 2: Check if a Number is Prime");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Prime number
        int num6 = 7;
        boolean result1 = isPrime(num6);
        System.out.println("Number: " + num6);
        System.out.println("Is Prime: " + result1);
        
        // Test Case 2 - Non-prime number
        int num7 = 10;
        boolean result2 = isPrime(num7);
        System.out.println("\nNumber: " + num7);
        System.out.println("Is Prime: " + result2);
        
        // Test Case 3 - Smallest prime
        int num8 = 2;
        boolean result3 = isPrime(num8);
        System.out.println("\nNumber: " + num8);
        System.out.println("Is Prime: " + result3);
        
        // Test Case 4 - Number less than 2
        int num9 = 1;
        boolean result4 = isPrime(num9);
        System.out.println("\nNumber: " + num9);
        System.out.println("Is Prime: " + result4);
        
        // Test Case 5 - Large prime
        int num10 = 17;
        boolean result5 = isPrime(num10);
        System.out.println("\nNumber: " + num10);
        System.out.println("Is Prime: " + result5);
        
        // Test Case 6 - Even number (except 2)
        int num11 = 4;
        boolean result6 = isPrime(num11);
        System.out.println("\nNumber: " + num11);
        System.out.println("Is Prime: " + result6);
        
        // Test Case 7 - Large non-prime
        int num12 = 100;
        boolean result7 = isPrime(num12);
        System.out.println("\nNumber: " + num12);
        System.out.println("Is Prime: " + result7);
        
        // Test Case 8 - Prime number
        int num13 = 13;
        boolean result8 = isPrime(num13);
        System.out.println("\nNumber: " + num13);
        System.out.println("Is Prime: " + result8);
    }
}

File: basic_dsa_five.py
"""
Basic DSA Problems - Two More Fundamental Questions
Simple and essential DSA problems for beginners
"""


# Problem 1: Find Factorial of a Number
# Calculate factorial using both iterative and recursive approaches

def factorial_iterative(n):
    """
    Iterative approach to find factorial
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if n < 0:
        return None  # Factorial is not defined for negative numbers
    
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result


def factorial_recursive(n):
    """
    Recursive approach to find factorial
    Time Complexity: O(n)
    Space Complexity: O(n) due to recursion stack
    """
    if n < 0:
        return None  # Factorial is not defined for negative numbers
    if n == 0 or n == 1:
        return 1
    return n * factorial_recursive(n - 1)


# Problem 2: Check if a Number is Prime
# Determine whether a number is prime or not

def is_prime(n):
    """
    Check if a number is prime
    Time Complexity: O(√n)
    Space Complexity: O(1)
    """
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    # Check divisibility from 3 to √n
    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2  # Only check odd numbers
    
    return True


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem 1: Find Factorial of a Number")
    print("=" * 60)
    
    # Test Case 1
    num1 = 5
    fact_iter1 = factorial_iterative(num1)
    fact_rec1 = factorial_recursive(num1)
    print(f"Number: {num1}")
    print(f"Factorial (Iterative): {fact_iter1}")
    print(f"Factorial (Recursive): {fact_rec1}")
    
    # Test Case 2
    num2 = 0
    fact_iter2 = factorial_iterative(num2)
    fact_rec2 = factorial_recursive(num2)
    print(f"\nNumber: {num2}")
    print(f"Factorial (Iterative): {fact_iter2}")
    print(f"Factorial (Recursive): {fact_rec2}")
    
    # Test Case 3
    num3 = 1
    fact_iter3 = factorial_iterative(num3)
    fact_rec3 = factorial_recursive(num3)
    print(f"\nNumber: {num3}")
    print(f"Factorial (Iterative): {fact_iter3}")
    print(f"Factorial (Recursive): {fact_rec3}")
    
    # Test Case 4
    num4 = 7
    fact_iter4 = factorial_iterative(num4)
    fact_rec4 = factorial_recursive(num4)
    print(f"\nNumber: {num4}")
    print(f"Factorial (Iterative): {fact_iter4}")
    print(f"Factorial (Recursive): {fact_rec4}")
    
    # Test Case 5 - Negative number
    num5 = -5
    fact_iter5 = factorial_iterative(num5)
    fact_rec5 = factorial_recursive(num5)
    print(f"\nNumber: {num5}")
    print(f"Factorial (Iterative): {fact_iter5}")
    print(f"Factorial (Recursive): {fact_rec5}")
    
    print("\n" + "=" * 60)
    print("Problem 2: Check if a Number is Prime")
    print("=" * 60)
    
    # Test Case 1 - Prime number
    num6 = 7
    result1 = is_prime(num6)
    print(f"Number: {num6}")
    print(f"Is Prime: {result1}")
    
    # Test Case 2 - Non-prime number
    num7 = 10
    result2 = is_prime(num7)
    print(f"\nNumber: {num7}")
    print(f"Is Prime: {result2}")
    
    # Test Case 3 - Smallest prime
    num8 = 2
    result3 = is_prime(num8)
    print(f"\nNumber: {num8}")
    print(f"Is Prime: {result3}")
    
    # Test Case 4 - Number less than 2
    num9 = 1
    result4 = is_prime(num9)
    print(f"\nNumber: {num9}")
    print(f"Is Prime: {result4}")
    
    # Test Case 5 - Large prime
    num10 = 17
    result5 = is_prime(num10)
    print(f"\nNumber: {num10}")
    print(f"Is Prime: {result5}")
    
    # Test Case 6 - Even number (except 2)
    num11 = 4
    result6 = is_prime(num11)
    print(f"\nNumber: {num11}")
    print(f"Is Prime: {result6}")
    
    # Test Case 7 - Large non-prime
    num12 = 100
    result7 = is_prime(num12)
    print(f"\nNumber: {num12}")
    print(f"Is Prime: {result7}")
    
    # Test Case 8 - Prime number
    num13 = 13
    result8 = is_prime(num13)
    print(f"\nNumber: {num13}")
    print(f"Is Prime: {result8}")
